bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java

```diff
@@ -651,6 +664,61 @@ private Request sendRequestBuilder(String subUrl, HttpMethod method) {
                 .agent(TAHOMA_AGENT);
     }
 
+    /**
+     * Performs the login for Cozytouch using OAUTH2 authorization.
+     *
+     * @return JSESSION ID cookie value.
+     * @throws ExecutionException
+     * @throws TimeoutException
+     * @throws InterruptedException
+     */
+    private String loginCozytouch()
+            throws InterruptedException, TimeoutException, ExecutionException, JsonSyntaxException {
+        String authBaseUrl = "https://" + COZYTOUCH_OAUTH2_URL;
+
+        String urlParameters = "grant_type=password&username=" + urlEncode(thingConfig.getEmail()) + "&password="
+                + urlEncode(thingConfig.getPassword());
+
+        ContentResponse response = httpClient.newRequest(authBaseUrl + COZYTOUCH_OAUTH2_TOKEN_URL)
+                .method(HttpMethod.POST).header(HttpHeader.ACCEPT_LANGUAGE, "en-US,en")
+                .header(HttpHeader.ACCEPT_ENCODING, "gzip, deflate").header("X-Requested-With", "XMLHttpRequest")
+                .header(HttpHeader.AUTHORIZATION, "Basic " + COZYTOUCH_OAUTH2_BASICAUTH)
+                .timeout(TAHOMA_TIMEOUT, TimeUnit.SECONDS).agent(TAHOMA_AGENT)
+                .content(new StringContentProvider(urlParameters), "application/x-www-form-urlencoded; charset=UTF-8")
+                .send();
+
+        if (response.getStatus() != 200) {
+            // Login error
+            if (response.getHeaders().getField(HttpHeader.CONTENT_TYPE).getValue()
+                    .equalsIgnoreCase(MediaType.APPLICATION_JSON)) {
+                try {
+                    SomfyTahomaOauth2Error error = gson.fromJson(response.getContentAsString(),
+                            SomfyTahomaOauth2Error.class);
+                    throw new ExecutionException(error.getErrorDescription(), null);
+                } catch (JsonSyntaxException e) {
+
+                }
+            }
+            throw new ExecutionException("Unknown error while attempting to log in.", null);
+        }
+
+        SomfyTahomaOauth2Reponse oauth2response = gson.fromJson(response.getContentAsString(),
+                SomfyTahomaOauth2Reponse.class);
+
+        logger.debug("OAuth2 Access Token: {}", oauth2response.getAccessToken());
+
+        response = httpClient.newRequest(authBaseUrl + COZYTOUCH_OAUTH2_JWT_URL).method(HttpMethod.GET)
+                .header(HttpHeader.AUTHORIZATION, "Bearer " + oauth2response.getAccessToken())
+                .timeout(TAHOMA_TIMEOUT, TimeUnit.SECONDS).send();
+
+        if (response.getStatus() == 200) {
+            String jwt = response.getContentAsString();
+            return jwt.replace("\"", "");
+        } else {
+            throw new ExecutionException(response.getContentAsString(), null);
```

Comment:
Just a thought: are we sure that the content of the response will be a good exception message ?!
Suggestion: Maybe a simple message could be better, something like: Failed to retrieve the JSESSION ID using the access token.

Reply:
I added the response code et response content as string to the exception message.